diff --git a/test/Payment.php b/test/Payment.php
index d711b80..27ccabb 100644
--- a/test/Payment.php
+++ b/test/Payment.php
@@ -134,8 +134,8 @@ class Payment extends \Magento\Payment\Model\Method\Cc
         $this->is_sandbox = $this->getConfigData('is_sandbox');
         $this->country = $this->getConfigData('country');
 
-        $this->_canRefund = $this->country === 'MX' ? true : false;
-        $this->_canRefundInvoicePartial = $this->country === 'MX' ? true : false;
+        //$this->_canRefund = $this->country === 'MX' ? true : false;
+        //$this->_canRefundInvoicePartial = $this->country === 'MX' ? true : false;
         
         $this->sandbox_merchant_id = $this->getConfigData('sandbox_merchant_id');
         $this->sandbox_sk = $this->getConfigData('sandbox_sk');
