diff --git a/final_price.phtml b/final_price.phtml
index 1577ddc..78b5e1d 100644
--- a/final_price.phtml
+++ b/final_price.phtml
@@ -23,8 +23,8 @@ $schema = ($block->getZone() == 'item_view') ? true : false;
 ?>
 </span>
 
-<?php if (!$block->isProductList() && $block->hasSpecialPrice()) : ?>
-    <span class="old-price sly-old-price no-display">
+<?php if ($block->hasSpecialPrice()) : ?>
+    <span class="old-price sly-old-price">
         <?= /* @noEscape */ $block->renderAmount($priceModel->getAmount(), [
             'display_label'     => __('Regular Price'),
             'price_id'          => $block->getPriceId('old-price-' . $idSuffix),
