diff --git a/Model/Address/Validator/General.php b/Model/Address/Validator/General.php
index 7cbb6ef..348a3ef 100644
--- a/Model/Address/Validator/General.php
+++ b/Model/Address/Validator/General.php
@@ -5,6 +5,7 @@
  */
 namespace Magento\Customer\Model\Address\Validator;
 
+use Magento\Customer\Api\AddressMetadataInterface;
 use Magento\Customer\Model\Address\AbstractAddress;
 use Magento\Customer\Model\Address\ValidatorInterface;
 
@@ -87,6 +88,7 @@ class General implements ValidatorInterface
      */
     private function checkOptionalFields(AbstractAddress $address)
     {
+		$this->reloadAddressAttributes($address);
         $errors = [];
         if ($this->isTelephoneRequired()
             && !\Zend_Validate::is($address->getTelephone(), 'NotEmpty')
@@ -148,4 +150,17 @@ class General implements ValidatorInterface
     {
         return $this->eavConfig->getAttribute('customer_address', 'fax')->getIsRequired();
     }
+	
+	/**
+     * Reload address attributes for the certain store
+     *
+     * @param AbstractAddress $address
+     * @return void
+     */
+    private function reloadAddressAttributes(AbstractAddress $address): void
+    {
+        $attributeSetId = $address->getAttributeSetId() ?: AddressMetadataInterface::ATTRIBUTE_SET_ID_ADDRESS;
+        $address->setData('attribute_set_id', $attributeSetId);
+        $this->eavConfig->getEntityAttributes(AddressMetadataInterface::ENTITY_TYPE_ADDRESS, $address);
+    }
 }
